numpy/random/_generator.pyx

```diff
@@ -4150,17 +4156,33 @@ cdef class Generator:
 
         i = 0
         totsize = np.PyArray_SIZE(val_arr)
-        with self.lock, nogil:
-            while i < totsize:
-                acc = 0.0
-                for j in range(k):
-                    val_data[i+j] = random_standard_gamma(&self._bitgen,
-                                                              alpha_data[j])
-                    acc = acc + val_data[i + j]
-                invacc = 1/acc
-                for j in range(k):
-                    val_data[i + j] = val_data[i + j] * invacc
-                i = i + k
+
+        # Small alpha case: Use stick-breaking approach with beta random variates
+        if alpha_arr.max() < 1.0:
+            with self.lock, nogil:
+                while i < totsize:
+                    acc = 1.
+                    for j in range(k-1):
+                        val_data[i+j] = acc * random_beta(&self._bitgen,
+                                                          alpha_data[j],
+                                                          alpha_csum_data[j+1])
+                        acc *= (1. - val_data[i+j])
```

Comment:
It looks like you want `acc` to "accumulate" the products of 1 minus the beta random variates, but you can't use `val_data[i+j]` for that, because it includes the product of `acc` and the random beta (not just the random beta).  You can make this work if you save the random beta variate in its own variable.  The body of the `while` loop then looks like this:

                    acc = 1.
                    for j in range(k - 1):
                        v = random_beta(&self._bitgen,
                                        alpha_data[j], alpha_csum_data[j + 1])
                        val_data[i + j] = acc * v
                        acc *= (1. - v)
                    val_data[i + j + 1] = acc
                    i = i + k

where `v` is `cdef`'d to be type `double` at the beginning of the function.

Reply:
Thanks for the hint. Will change as suggested.

Reply:
Thanks, the update looks good.

I'm currently running a bunch of tests to see if anything breaks. :)